### ISSUE TITLE
Deprecate `prepare:`

### ISSUE THREAD

**Issue by rmosolgo** [rmosolgo/graphql-ruby#2749]
Prepare was added a long time ago, but I've never used it and I'd like to remove it from GraphQL-Ruby because it's hard to make it work right. 

For example: 

- https://github.com/rmosolgo/graphql-ruby/issues/982
- https://github.com/rmosolgo/graphql-ruby/issues/1024
- https://github.com/rmosolgo/graphql-ruby/issues/2001
- https://github.com/rmosolgo/graphql-ruby/issues/1641
- https://github.com/rmosolgo/graphql-ruby/issues/2721

The problems are: 

- Some `prepare:` implementations (namely, methods) assume that an underlying object is present, but this isn't always the case (as during `subscriptions.trigger(...)`) 
- Similarly for `context` -- it's not present during `subscriptions.trigger`, but it's part of the `prepare:` API 
- Prepare was supposed to _always_ convert a "raw" incoming value to a "prepared" one, but this is confusing to people, because `prepare:` runs _before_ the query. It's executed early because analysis needs the prepared value. Extra confusing, `prepare:` for mutations is non-sequential. 

So, my question is: 

- What alternative approach could replace `prepare:`?
- Are there any use cases of `prepare:` that couldn't be replaced with an existing feature? 
- Similarly, is there any compelling reason _not_ to remove `prepare:`, except that it's already widely used (despite the surprising behavior and bugs) 

cc @cainlevy, @swalkinshaw , @modosc  whose issues are linked above. Do you think it can/should be fixed? Or preserved as-is? Or replaced/removed?

**Comment by swalkinshaw** [comment]
My main question is why isn't (or can't) `prepare` just be called during execution at runtime?

I'm wondering if those bugs linked actually still happen with 1.10... since from what I can see `prepare_value` is only called in `interpreter/runtime`. Actually `Schema::InputObject` still calls it during its `initialize` though I'm not familiar enough with *when* those are initialized.

**Comment by rmosolgo** [comment]
> just be called during execution

Y'know, maybe it could. The previous reason is, _analysis_ had access to field arguments, and `prepare` applied to field arguments, therefore `prepare` was applied during analysis (sometimes). 

But the new analysis module is AST-based, so maybe we could roll back that behavior in a reasonable way. 

But _some_ argument-related preparation should probably still happen during analysis. For example, variable values should probably still be merged in, and argument default values should probably be applied as needed. Do you agree? 

In that case, we'd separate argument preparation into two steps: 

- Query transformation: add variable values and default values to the arguments hash (validation/analysis)
- Prepare: call `prepare:` hooks on members of the arguments hash (runtime)

**Comment by swalkinshaw** [comment]
Yes I believe that make sense 👍 When I was talking about `prepare` I specifically meant that user-exposed hook; not the generic "argument preparation".

I assume the first point already happens and it's separate from the `prepare` hook? 

Our use case which we'd selfishly like to keep is validating + converting IDs. So our arguments in resolvers are converted to an ID object instead of just the string value. We'd be open to any equivalent to that of course but I do think `prepare` is valuable for that kind of use case.

**Comment by JanStevens** [comment]
Our use case is to mutate the given argument input. For example we have some search fields and the frontend sends the value as: `CLIENT_ADMIN` while we expect for the search to be `ClientAdmin`.

TBH we could mutate the input object in the `resolve` method but `prepare` seemed a bit cleaner and allows to reuse those input objects

```ruby
prepare: ->(scope, _ctx) { scope.to_s.classify }
```

**Comment by rmosolgo** [comment]
Thanks for sharing that example, @JanStevens. Is `CLIENT_ADMIN` an enum value? If so, another option is to use `value:` , eg 

```
class Types::SearchField < Types::BaseEnum
  # ...
  # `CLIENT_ADMIN` is sent and received from GraphQL; `ClientAdmin` is passed into Ruby code
  value "CLIENT_ADMIN", value: ClientAdmin
end 
```

Would that work in your case? (I'm not suggesting you change it, I'm just interested to learn more about the situation.)

**Comment by JanStevens** [comment]
> Thanks for sharing that example, @JanStevens. Is `CLIENT_ADMIN` an enum value? If so, another option is to use `value:` , eg
> 
> ```
> class Types::SearchField < Types::BaseEnum
>   # ...
>   # `CLIENT_ADMIN` is sent and received from GraphQL; `ClientAdmin` is passed into Ruby code
>   value "CLIENT_ADMIN", value: ClientAdmin
> end 
> ```
> 
> Would that work in your case? (I'm not suggesting you change it, I'm just interested to learn more about the situation.)

Sadly not really, to expand on the use case, the places where we use it are for inputs that deal with searching. We use the ransack search gem so we need to convert `ClientAdmin` to a class that can be used in a polymorphic relation search `where(admin_type: 'ClientAdmin')` but in the enum case we do want to have it as `client_admin` since this allows us to load in the right "role" file.

Ofcourse we could update the loading of the right role file but I would like to keep my enum's consistent

**Comment by jturkel** [comment]
We use argument `prepare` hooks to declaratively define common argument validations that can't be expressed via the GraphQL type system. For example the following defines a `percentage` argument that must be between 0 and 1:

```ruby
class Image < BaseObject
  field :scaled_content, URL, null: false do
    argument :percentage, Float, required: true, min: 0, max: 1
  end
end
```
We probably could have implemented it via field extensions but it was easier with argument prepare hooks.

**Comment by modosc** [comment]
@jturkel thats nice! we’ve solved those problems with analyzers instead but the require a bit more code to achieve the same thing.

**Comment by LucDelmon** [comment]
Hi,

I see too case were prepare is useful for me.

-  On data manipulation, especially on dates when we want to 'prepare' it for the rest of the query
```
argument :starts_at, GraphQL::Types::ISO8601DateTime, required: true, prepare: lambda { |starts_at, _ctx|
      starts_at.in_time_zone.beginning_of_day
    }
```
but it can definitely be done later in the code

- On array of ids. `argument :user_ids, [ID]`. Using loads: call `Schema.Object_from_id` for every element. With prepare I can retrieve them in only one.

```
argument :user_ids, [ID], required: true, prepare: lambda { |ids, _ctx|
      User.in(id: ids).to_a
    }, as: :users
```

I don't see an alternative but would love to be able to surcharge the loads logic on array.

**Comment by rmosolgo** [comment]
Thanks for sharing those examples, that's really helpful 🍻 !

**Comment by lifeiscontent** [comment]
@rmosolgo I'd love to see an API change to how arguments work basically a similar interface as fields in a type

```rb
# frozen_string_literal: true

module Mutations
  class UpdateUser < Mutations::BaseMutation
    class UpdateUserInput < Types::BaseInputObject
      argument :first_name, String, required: false, default_value: nil
      def first_name(value, context)
        return value.upcase if value.present?

        value
      end

      def prepare
        to_h
      end
    end

    argument :id, ID, required: true
    argument :input, UpdateNurseProfileInput, required: true

    field :user, Types::UserType, null: false

    def resolve(id: nil, input:)
      user = User.find_by!(uuid: id)

      authorize! user, to: :update?

      user.update!(input)

      {
        user: user.reload
      }
    end
  end
end
```

I don't have the context of everything that's happening in the resolver, but in order to get a consistent value back during an argument transform, this is what makes sense to me.

**Comment by digitalmoksha** [comment]
@rmosolgo I'm wondering if you're still planning/considering deprecating `prepare:`.  We're just starting to use it in a couple places, such as massaging an id for backward compatibility.  But if it is going to be deprecated, it's better for us to use a different method.  For example, here's one use:

```ruby
             prepare: ->(id, _ctx) do
               # Originally accepted a raw model id. Now accept a gid, but allow a raw id
               # for backward compatibility
               GitlabSchema.parse_gid(id, expected_type: ::Iteration).model_id
             rescue Gitlab::Graphql::Errors::ArgumentError
               id
             end
```

**Comment by rmosolgo** [comment]
I'm not sure, sorry. I've never used it, and it's a pain to maintain, but lots of folks really value it (as shown here). So maybe my interest in simplifying the code is outweighed by the existing (and future) uses of it. Whatever happens, it'll have a clear migration path with support for whatever it can already do. So, I think it's :+1: to use.

**Comment by digitalmoksha** [comment]
> Whatever happens, it'll have a clear migration path with support for whatever it can already do. So, I think it's 👍 to use.

Awesome, thanks!
